core/src/protocols_handler/node_handler.rs

```diff
@@ -224,8 +224,15 @@ where
                     self.negotiating_out.push((upgr_info, in_progress));
                 }
                 Err(err) => {
-                    let msg = format!("Error while upgrading: {:?}", err);
-                    let err = io::Error::new(io::ErrorKind::Other, msg);
+                    let err = if err.is_elapsed() {
+                        let err = err.into_timer().expect("is_elapsed returned true; QED");
```

Comment:
This is wrong. `into_timer` only returns `Some` if the timer implementation itself errors. An action timing out is different and expected, in other words, if `is_elapsed` is true there is no inner error to return.